jobs: OnFailOrCancel is now resumable

spaskob opened the pull request:
OnFailOrCancel is a post job execution hook
that is used to revert any changes left from a
failed or canceled jobs. So far this function
was not guaranteed to complete if the node running
the job fails or if the job is interrupted for
other reasons. We fix that by treating OnFailOrCancel
similarly to Resume. ie it will be executed when a job
is adopted and is in state Reverting.

In addition to the new job state Reverting that job
goes in when it's Resume function fails or when the job
is canceled we add state CancelRequested to indicate that
the user has requested the job to be canceled but job is still
in Resume. When the request is detected the job will transition
to Reverting state. This means that now we should be adopting not
only running jobs but also cancel-requested and reverting.

Additional changes:
- simplify the handling of nil closures when updating a job status
- move stopper signal to quiesce first in select statements

Fixes #41572.

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44163)
<!-- Reviewable:end -->

Commit:
jobs: OnFailOrCancel is now resumable

OnFailOrCancel is a post job execution hook
that is used to revert any changes left from a
failed or canceled jobs. So far this function
was not guranteed to complete if the node running
the job fails or if the job is interrupted for
other reasons. We fix that by treating OnFailOrCancel
similarly to REsume. ie it will beexecuted when a job
is adopted and is in state Reverting.

In additioan to the new job state Reverting that job
goes in when it's Resume function fails or when the job
is canceled we add state CancelRequested to indicate that
the user has requested the job to be cnaceld but job is still
in Resume. When the request is detected the job will transition
to Reverting state. This means that now we should be adopting not
only running jobs but also cancel-requested and reverting.

Additional changes:
- simplify the handling of nil closures when updating a job status
- move stopper signal to quiesce firts in selec statements

Release note: none

thoszhang reviewed (commented):
> Reviewed 14 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @spaskob)

___
*[pkg/jobs/jobs.go, line 332 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhUb1i0cNWWmvLgYmY:-LzhUb1i0cNWWmvLgYmZ:b6ci3yo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/jobs.go#L332)):*
> ```Go
> // that it is in state StatusCancelRequested and will move it to state
> // StatusReverting.
> func (j *Job) cancelRequested(
> ```

Why do we export `Reverted` but not `cancelRequested`?
___
*[pkg/jobs/jobs.go, line 336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhSl265XbGAdt6T-VK:-LzhSl265XbGAdt6T-VL:b3cyf9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/jobs.go#L336)):*
> ```Go
> ) error {
> 	return j.Update(ctx, func(txn *client.Txn, md JobMetadata, ju *JobUpdater) error {
> 		if md.Status == StatusCancelRequested {
> ```

What if the status is `StatusReverting`?
___
*[pkg/jobs/jobs.go, line 348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhSJz506p6y4ogWPfP:-LzhSJz506p6y4ogWPfQ:bfvxx4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/jobs.go#L348)):*
> ```Go
> 		}
> 		ju.UpdateStatus(StatusCancelRequested)
> 		// TODO(spaskob): it is not clear why payload needs to be updated.
> ```

Does it? I thought previously in e.g., `Job.canceled()` we were updating `Payload.FinishedMicros`, but now that we're not updating the payload fields, there's no reason. 
___
*[pkg/jobs/jobs.go, line 401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhTW_m0XzaxesiisR0:-LzhTW_m0XzaxesiisR1:b1lh76l) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/jobs.go#L401)):*
> ```Go
> ) error {
> 	return j.Update(ctx, func(txn *client.Txn, md JobMetadata, ju *JobUpdater) error {
> 		// TODO(spaskob): should we fail if the terminal state is not StatusFailed?
> ```

Do we generally expect callers to check the job state before trying to update it? If not, then we probably shouldn't fail.

In general I wish there were a more systematic way to enforce allowed/disallowed state transitions, aside from having a lot of if statements at the top of each attempted state transition, but that's outside the scope of this change.
___
*[pkg/jobs/registry.go, line 639 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhXgSIAQIiadkz3w92:-LzhXgSJ-pqjTyFNO4OO:by5gseg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/registry.go#L639)):*
> ```Go
> 	jobErr error,
> ) error {
> 	log.Infof(ctx, "job %d: stepping through state %s with error %v", *job.ID(), status, jobErr)
> ```

nit: I think this should just be `job %d: stepping through state %s` if there's no error, otherwise the logs will be confusing.
___
*[pkg/jobs/registry.go, line 918 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-Lzh_Udf73Jo5S_u-v2s:-Lzh_Udf73Jo5S_u-v2t:b-4iojff) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/registry.go#L918)):*
> ```Go
> 					log.Infof(ctx, "job %d: canceling: the job is cancelRequested and running on this node", *id)
> 				}
> 				r.unregister(*id)
> ```

Why do we unregister here?


<!-- Sent from Reviewable.io -->

Commit:
jobs: restrict scope of pause and cancel

Pause is allowed only for pending and running jobs.
Cancel s only allowed for pending, running or paused jobs.
A job can be marked canceled only if it is in reverting state.

Release note (<category, see below>): <what> <show> <why>

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @lucy-zhang)

___
*[pkg/jobs/jobs.go, line 332 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhUb1i0cNWWmvLgYmY:-LzhdRWW76II0HD_uyfQ:b-ft1xgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/jobs.go#L332)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Why do we export `Reverted` but not `cancelRequested`?
</blockquote></details>

good question, I think all should not be exported, but for example Resume is also exported. So the answer is I don't know!
___
*[pkg/jobs/jobs.go, line 336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhSl265XbGAdt6T-VK:-LzhdgOdAHE3OggZnFAH:b-erblis) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/jobs.go#L336)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

What if the status is `StatusReverting`?
</blockquote></details>

as per the slack convo, we decided that reverting is not cancellable but we will make it pausable in the future.
___
*[pkg/jobs/jobs.go, line 348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhSJz506p6y4ogWPfP:-LzhdsBFBA51_ZZ2nCj9:b-frgrah) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/jobs.go#L348)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Does it? I thought previously in e.g., `Job.canceled()` we were updating `Payload.FinishedMicros`, but now that we're not updating the payload fields, there's no reason. 
</blockquote></details>

that's my guess as well, I will remove this after making sure there are no regressions.
___
*[pkg/jobs/jobs.go, line 401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhTW_m0XzaxesiisR0:-LzheBMU5MRA275-eNU9:b-40xfz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/jobs.go#L401)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Do we generally expect callers to check the job state before trying to update it? If not, then we probably shouldn't fail.

In general I wish there were a more systematic way to enforce allowed/disallowed state transitions, aside from having a lot of if statements at the top of each attempted state transition, but that's outside the scope of this change.
</blockquote></details>

I agree, this is inherited complexity from jobs infrastructure - but it has been quite fragile to modify. So I am punting on most of these and adding TODOs as reminders for future more fundamental refactorings/re-designs.
___
*[pkg/jobs/registry.go, line 639 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhXgSIAQIiadkz3w92:-LzhegjlA8osn8C9ec68:bpn5c0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/registry.go#L639)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: I think this should just be `job %d: stepping through state %s` if there's no error, otherwise the logs will be confusing.
</blockquote></details>

the log message is with error <nil> which is ok imo, it's a balance between adding more boilerplate code.
___
*[pkg/jobs/registry.go, line 918 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-Lzh_Udf73Jo5S_u-v2s:-Lzhes9g6K711uRwWZt5:bncv0kx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/registry.go#L918)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Why do we unregister here?
</blockquote></details>

the explanation is in the log message above, lmk if it is not clear. The job has been requested to be canceled and this node is running it so we unregister it (which will stop its execution). Before we relied on the job noticing that it is canceled.


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @lucy-zhang)

___
*[pkg/jobs/jobs.go, line 348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhSJz506p6y4ogWPfP:-Lzhftjt4XIDKCcxl_Xf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/jobs.go#L348)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

that's my guess as well, I will remove this after making sure there are no regressions.
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 639 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhXgSIAQIiadkz3w92:-LzhfzXI1Jf9-jE6n3OT:bsatwis) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/registry.go#L639)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

the log message is with error <nil> which is ok imo, it's a balance between adding more boilerplate code.
</blockquote></details>

it looks like "... with error <nil>"


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> Reviewed 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @lucy-zhang, and @spaskob)

___
*[pkg/jobs/jobs.go, line 336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhSl265XbGAdt6T-VK:-LzmrBBbBn2O-lJY9PAn:b9ny26) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/jobs.go#L336)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

as per the slack convo, we decided that reverting is not cancellable but we will make it pausable in the future.
</blockquote></details>

`StatusReverting` is not a terminal state, right? Don't we have to handle it separately?
___
*[pkg/jobs/registry.go, line 639 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhXgSIAQIiadkz3w92:-LzmrUFE7beljAfooAZU:b91vicb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/registry.go#L639)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

it looks like "... with error <nil>"
</blockquote></details>

Oh, since we're using %v a nil error gets printed as `<nil>`. That's fine, then.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> Reviewed 2 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @lucy-zhang, and @spaskob)

___
*[pkg/jobs/jobs.go, line 336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44163#-LzhSl265XbGAdt6T-VK:-LzmtDJB9sdJadCl_HWX:b-9pt6wk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b59b19ce10220bb9ee2c93e6cde6ec866f6d3ab/pkg/jobs/jobs.go#L336)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

`StatusReverting` is not a terminal state, right? Don't we have to handle it separately?
</blockquote></details>

I just saw the second commit, nvm


<!-- Sent from Reviewable.io -->

thoszhang reviewed (approved):
> Reviewed 1 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @spaskob)



<!-- Sent from Reviewable.io -->

spaskob commented:
bors r+

spaskob commented:
bors r+
